### ISSUE TITLE
dbatools module import errors

### ISSUE THREAD

**Issue by ACALVETT** [sqlcollaborative/dbatools#4965]
[importModuleOutputs.txt](https://github.com/sqlcollaborative/dbatools/files/2764584/importModuleOutputs.txt)

## Steps to Reproduce

$dbatools_serialimport=$true
import-module  dbatools -Verbose

## Expected Behavior

dbatools module should load

## Actual Behavior

Using the latest build of dbatools and importing the module i get

WARNING: The module import has hit a timeout while waiting for some background tasks to finish.
This may result in some commands not working as intended.
This should not happen under reasonable circumstances, please file an issue at:
https://github.com/sqlcollaborative/dbatools/issues
Or contact us directly in the #dbatools channel of the SQL Server Community Slack Channel:
https://dbatools.io/slack/
Timeout waiting for temporary runspaces reached! The Module import will complete, but some things may not work as intended

If i then unload the module and add $dbatools_serialimport=$true i get 
```
Exception calling "Invoke" with "3" argument(s): "Cannot bind argument to parameter 'Path' because it is null."
At line:203 char:5
+     $scriptBlock.Invoke($script:PSModuleRoot, "$script:DllRoot\smo",  ...
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : CmdletInvocationException
```

The module did load once during repo but i cant explain why but it now consistently fails
```
## Environmental data

Name                           Value                                                                                                                                                                                                                             
----                           -----                                                                                                                                                                                                                             
PSVersion                      5.1.14409.1018                                                                                                                                                                                                                    
PSEdition                      Desktop                                                                                                                                                                                                                           
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                                                                                                                                                           
BuildVersion                   10.0.14409.1018                                                                                                                                                                                                                   
CLRVersion                     4.0.30319.42000                                                                                                                                                                                                                   
WSManStackVersion              3.0                                                                                                                                                                                                                               
PSRemotingProtocolVersion      2.3                                                                                                                                                                                                                               
SerializationVersion           1.1.0.1      
```

**Comment by ACALVETT** [comment]
Scope ExecutionPolicy
        ----- ---------------
MachinePolicy    RemoteSigned
   UserPolicy    RemoteSigned
      Process       Undefined
  CurrentUser          Undefined
 LocalMachine       Undefined

**Comment by wsmelton** [comment]
Do you have any security software or endpoint protection running on your machine?

**Comment by ACALVETT** [comment]
McAfee & FireEye endpoint are ones i know are running and i can not stop or configure them

**Comment by potatoqualitee** [comment]
great points, thank you shawn. i can confirm that mcafee works with no issue but i am unsure about fireeye.

**Comment by wsmelton** [comment]
McAfee works in your environment with no issues @potatoqualitee , that won't apply to everyone because it is based on what policies have been pushed out to the client agents. The Access Protection is the biggest issue with that product.

@ACALVETT just a shot in the dark can you try starting powershell.exe in this manner (just paste it into a run prompt):

```
powershell -noprofile -ExecutionPolicy bypass
```

Then try to import the  module and see if you get the same results. Then if you will collect the output from this `[Sqlcollaborative.Dbatools.dbaSystem.DebugHost]::ImportTime` and post it to the issue.

**Comment by wsmelton** [comment]
Actually scratch that last one we apparently have a command for collecting import data: `Measure-DbatoolsImport`

**Comment by ACALVETT** [comment]
WARNING: The module import has hit a timeout while waiting for some background tasks to finish.
This may result in some commands not working as intended.
This should not happen under reasonable circumstances, please file an issue at:
https://github.com/sqlcollaborative/dbatools/issues
Or contact us directly in the #dbatools channel of the SQL Server Community Slack Channel:
https://dbatools.io/slack/
Timeout waiting for temporary runspaces reached! The Module import will complete, but some things may not work as intended

Action                                 Duration         Start               End                
------                                 --------         -----               ---                
Loading import helper functions        00:00:00.0260026 16/01/2019 15:17:36 16/01/2019 15:17:36
Resolved path to not SQLSERVER PSDrive 00:00:00.0070007 16/01/2019 15:17:36 16/01/2019 15:17:36
Validated defines                      00:00:00.0750075 16/01/2019 15:17:36 16/01/2019 15:17:36
Unblocking Files                       00:00:01.7241724 16/01/2019 15:17:36 16/01/2019 15:17:38
Starting import SMO libraries          00:00:00.1110111 16/01/2019 15:17:38 16/01/2019 15:17:38
Loading dbatools library               00:00:00.6850685 16/01/2019 15:17:38 16/01/2019 15:17:39
Loading Public and Private Commands    00:00:00.7140714 16/01/2019 15:17:39 16/01/2019 15:17:40
Registering cmdlets                    00:00:00.0770077 16/01/2019 15:17:40 16/01/2019 15:17:40
Configuration System                   00:00:00.8770877 16/01/2019 15:17:40 16/01/2019 15:17:41
Loading Optional Commands              00:00:00.3620362 16/01/2019 15:17:41 16/01/2019 15:17:41
Loading TEPP                           00:00:00.2480248 16/01/2019 15:17:41 16/01/2019 15:17:41
Loading Message Transforms             00:00:00.0560056 16/01/2019 15:17:41 16/01/2019 15:17:41
Script: Logging                        00:00:00.1110111 16/01/2019 15:17:41 16/01/2019 15:17:41
Script: Asynchronous TEPP Cache        00:00:00.0770077 16/01/2019 15:17:41 16/01/2019 15:17:41
Script: Maintenance                    00:00:00.2170217 16/01/2019 15:17:41 16/01/2019 15:17:42
Loading Aliases                        00:00:00.0620062 16/01/2019 15:17:42 16/01/2019 15:17:42
Exported module member                 00:00:00.1530153 16/01/2019 15:17:42 16/01/2019 15:17:42
Waiting for runspaces to finish        00:00:24.9894987 16/01/2019 15:17:42 16/01/2019 15:18:07
Loaded type extensions                 00:00:00.5880588 16/01/2019 15:18:07 16/01/2019 15:18:07

**Comment by ACALVETT** [comment]
Just to be clear, thats the output having done the 

> powershell -noprofile -ExecutionPolicy bypass

**Comment by potatoqualitee** [comment]
Thanks much. I remember coding this and i couldnt figure out the issue now i see room for improvement. I'm gonna go with shawn's suggestion that it's likely the anti-virus, but no worries, we can fix it on our side, potentially.

If it's possible and appropriate in your environment, can you try this?

https://github.com/sqlcollaborative/dbatools/archive/dll.zip

I just renamed the .zip to .dll. That should not work but i'm curious.

After that, can you test this version?

https://www.powershellgallery.com/packages/dbatools/0.9.701

That one has the old .ps1 import method.

**Comment by ACALVETT** [comment]
Hi,

So i downloaded the dll.zip one, unzipped it and then tried to load it up i got

import-module C:\UBS\Dev\Tools\dbatools -Verbose
VERBOSE: Loading module from path 'C:\UBS\Dev\Tools\dbatools\dbatools.psd1'.
VERBOSE: Loading 'TypesToProcess' from path 'C:\UBS\Dev\Tools\dbatools\xml\dbatools.Types.ps1xml'.
VERBOSE: Loading 'FormatsToProcess' from path 'C:\UBS\Dev\Tools\dbatools\xml\dbatools.Format.ps1xml'.
import-module : The following error occurred while loading the extended type data file: , C:\UBS\Dev\Tools\dbatools\xml\dbatools.Types.ps1xml: The file was skipped because of the 
following validation exception: File C:\UBS\Dev\Tools\dbatools\xml\dbatools.Types.ps1xml cannot be loaded. The file C:\UBS\Dev\Tools\dbatools\xml\dbatools.Types.ps1xml is not digitally 
signed. You cannot run this script on the current system. For more information about running scripts and setting execution policy, see about_Execution_Policies at 
http://go.microsoft.com/fwlink/?LinkID=135170..
At line:2 char:1
+ import-module C:\UBS\Dev\Tools\dbatools -Verbose
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [Import-Module], RuntimeException
    + FullyQualifiedErrorId : FormatXmlUpdateException,Microsoft.PowerShell.Commands.ImportModuleCommand

**Comment by ACALVETT** [comment]
Hi, 

Just tried release .9.701 and hit the error

Exception calling "Invoke" with "3" argument(s): "Cannot bind argument to parameter 'Path' because it is null."
At line:203 char:5
+     $scriptBlock.Invoke($script:PSModuleRoot, "$script:DllRoot\smo",  ...
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : CmdletInvocationException

**Comment by ACALVETT** [comment]
Hi @potatoqualitee 

I did a lot of digging this evening which i really enjoyed and learnt loads even though i don't understand a hell of a lot of the powershell (yet!).

Firstly i do not believe this is a antivirus problem and below is what i found.

I identified your smoLibraryImport.ps1 script as the place to investigate and it turns out that the error is thrown by the Copy-Assembly function but i did not get as far as working out why it throws that error.

One thing i did do was comment out the function, i don't fully understand why it is there and for debugging purposes, why not!  

What i then found was that i got a different error for a file for all the others including the SMO assembly loaded, consistently, and i got run dba commands! (Yay)   

Bound to be problems due to the other error being throw and the error i get with this hack is

Exception calling "Invoke" with "3" argument(s): "Could not load file or assembly 'file:///C:\ubs\Dev\Tools\dbatools\bin\smo\Microsoft.SqlServer.BatchParser.dll' or one of its dependencies. An attempt was made to load a program with an incorrect format."
At line:209 char:5
+     $scriptBlock.Invoke($script:PSModuleRoot, "$script:DllRoot\smo",  ...
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : CmdletInvocationException
 

Hope this helps to work out what is causing the problems!

**Comment by ACALVETT** [comment]
Its an offline install so I put the module in that directory and then ran `import-module dbatools -verbose`

Regarding the last exception, the SMO library is loaded because I can run dbatools commands that depend on the smo library and previously they would fail due to missing SMO.

**Comment by wsmelton** [comment]
Do you have SQL Server tools installed on your machine?

If you don't use the import of library in the module then it is using the version related with the GAC on your machine.

**Comment by ACALVETT** [comment]
Yes I have SQL Server tools on my machine.

To clear the confusion, I am using the smoLibraryImport I just commented out one function in it when I was testing, I commented out the Copy-Assembly function. It still runs and executes all the Add-Type commands in my test and that's how the SMO library got loaded.

If there is anything else I can capture let me know and I will.

**Comment by wsmelton** [comment]
Ah, ok. That function is actually never called to do any work the way I read it. If docopy is not passed in, which in the psm1 it is not then all that function does is return nothing and does no action.

**Comment by potatoqualitee** [comment]
thanks. I have some work to do on this command - i touched it a bit when i was porting to core and had such strange issues. this is all useful information, thank you both 👍

**Comment by potatoqualitee** [comment]
Follow up from Slack

```
samgreene [4:27 PM]
Hi All.  I've been using dbatools on my ssis/job server for the past few months without issue.  I have jobs that use cmd job steps an run as a windows login proxy.  I updated dbatools to the latest version using update-module and am getting an error when running jobs.  The 'test-DbaConnection'   command was found in the module 'dbatools', but the module could not be   loaded. For more information, run 'Import-Module dbatools'.
Then I add an import to the script and get this error:  Cannot bind argument to parameter 'Path' because it is null
If I log into the server and run PS as the proxy account, I am able to import dbatools without error.
```

**Comment by potatoqualitee** [comment]
@FriedrichWeinmann - it's more than msbuild.exe. Our process isn't streamlined and has too many variables.

* I believe the copydll variable sometimes builds a dll when it shouldn't (it acts like we enabled alwaysbuidlibrary, I think), maybe this is the msbuild.exe thing you were talking about
* Debugging requires two variables to be set (dbatools_multifileimport and dbatools_doDotSource)
* Isn't dbatools_multifileimport  and dbatools_doDotSource the same?
* The path isn't set sometimes

Like, there's just so many different ways for it to build that I lost track along the way and i'm pretty sure I introduced some issues. I swear sometimes a variable just doesn't get set when it should. 

While you are in our import process, can you please make it enable debugging if a .git directory is detected? I was going to do it myself but @wsmelton said it was more complex than just setting dbatools_multifileimport and dbatools_doDotSource if it detects .git.

**Comment by Stuart-Moore** [comment]
And to be a pain, could you please add a definitive guide to forcing a dll compile pretty please.

Been going round in circles trying to get a new version of the history object to fix an issue and non of the suggested fixes seem to work. This could well just be stupidity, but a document would rule that one out as well :)

**Comment by FriedrichWeinmann** [comment]
Heya @Stuart-Moore ,

the definitive guide to the import options (at least where triggering them is concerned) is hidden here:

```
Get-Help about_dbatools_importoptions
```

The problem was, that until the release just now, automatic library building _was broken_ . Silently broken, to make it even better. So no, totally not your fault you couldn't make it work.
The self-building capability was broken when updating for PS Core, as we needed a feature unavailable in earlier C# versions, but the old msbuild is deprecated and - since msbuild failed silently - nobody noticed. May need to add a test...

**Comment by propellor** [comment]
Hi folks, still running into this issue when importing the module in a sql job step of cmdexec type.  The step is running as a windows proxy acct that is admin on the machine.  

The workaround at this point is to log into the server and run powershell as the proxy, import the module and leave the session open.

**Comment by potatoqualitee** [comment]
ditto, trying myself on code. can repro with $script:multiFileImport = $true; $script:serialImport = $true

**Comment by potatoqualitee** [comment]
k i fixed it but gonna give it another day of testing before publishing to gallery 👍

**Comment by propellor** [comment]
Great!  Thank you!!!

**Comment by ACALVETT** [comment]
Hi, 

I downloaded the latest version today and got the timeout error again so i ran

```
$dbatools_serialimport=$true 
import-module dbatools
```
I am not seeing the path is null issue so looks like you fixed! However i am getting a different error

> 
> Exception calling "Invoke" with "3" argument(s): "Could not load file or assembly 'file:///P:\Documents\WindowsPowerShell\MappedModules\dbatools\bin\smo\Microsoft.SqlServer.BatchParser.dll' or one of its dependencies. An attempt was made to load a program with an incorrect format."
> At line:151 char:5
> +     $scriptBlock.Invoke($script:PSModuleRoot, "$script:DllRoot\smo",  ...
> +     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>     + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
>     + FullyQualifiedErrorId : CmdletInvocationException

Not sure where to start digging into that one, i do have SQL management tools installed on my machine.

**Comment by wsmelton** [comment]
Please create a new issue.

**Comment by propellor** [comment]
> k i fixed it but gonna give it another day of testing before publishing to gallery 👍

Was it posted?

**Comment by potatoqualitee** [comment]
yes indeed. when you post the new issue, please include what you can from `ls env:`

**Comment by ACALVETT** [comment]
Sure
